lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1223,6 +1264,48 @@ def _url(self, unit_number) -> str:
             self.port,
         )
 
+    def _check_alert_rules(self) -> bool:
+        """Check alert rules using Loki API.
+
+        Returns: bool
+
+        Emits:
+            loki_push_api_alert_rules_ok: This event is emitted when there are no errors with
+                alert rules.
+            loki_push_api_alert_rules_error: This event is emitted when the Loki API returns an
+                error because it failed to list rule group.
+        """
+        url = "http://127.0.0.1:{}/loki/api/v1/rules".format(self.port)
+        req = request.Request(url)
+        try:
+            request.urlopen(req)
+        except HTTPError as e:
+            msg = e.msg  # type: ignore
+            if e.code == 404 and "no rule groups found" in msg:
+                logger.debug("Checking alert rules: No rule groups found")
+                self.on.loki_push_api_alert_rules_ok.emit()
+                return True
+
+            if e.code == 404 and "404 page not found" in msg:
+                logger.error("Checking alert rules: 404 page not found: %s", url)
+                self.on.loki_push_api_alert_rules_error.emit()
+                return False
+
+            message = "{} - {}".format(e.code, msg)
+            logger.error("Checking alert rules: %s", message)
+            self.on.loki_push_api_alert_rules_error.emit()
+            return False
+        except URLError as e:
+            logger.error("Checking alert rules: %s", e.reason)
+            self.on.loki_push_api_alert_rules_error.emit()
+            return False
+        else:
+            logger.debug("Checking alert rules: Ok")
+            self.on.loki_push_api_alert_rules_ok.emit()
+            return True
```

Comment:
While this is certainly good, it could be improved further. We could check that the rules actually match what is supposed to be there. What happens when you give Loki one good and one bad alert rule? I think this check would pass.

Reply:
I am okay merging with this as is, but just wanted to mention it could be improved.

Reply:
> What happens when you give Loki one good and one bad alert rule?

Querying the endpoint `/loki/api/v1/rules` will return a `400` with:

```failed to list rule group for user fake and namespace loki_c937c803-173d-4c1f-896d-03ae800a0c82_prom_prom_0_promtail-k8s_alert.rules: error parsing /loki/rules/fake/loki_c937c803-173d-4c1f-896d-03ae800a0c82_prom_prom_0_promtail-k8s_alert.rules: /loki/rules/fake/loki_c937c803-173d-4c1f-896d-03ae800a0c82_prom_prom_0_promtail-k8s_alert.rules: could not parse expression: parse error at line 0, col 11: syntax error: unexpected RANGE, expecting NUMBER or { or (```

As far as I understand when querying `/loki/api/v1/rules` Loki returns alert rules that are stored in `/loki/rules/fake` directory...  @simskij  Do you think it's worth checking again here?

Reply:
I'd imagine this responsibility belongs to loki and hopefully in the near future there will be api / web ui around invalid config details.

Reply:
Grafana is the web API, generally. If it changes in the future, I'd imagine that it will end up like the change to Grafana, which now returns a 200 OK along with a JSON message saying "actually it's not ok."

An integration test is important here.

Reply:
Integration test added.